api/epoch_based_provider.go

```diff
@@ -0,0 +1,204 @@
+package api
+
+import (
+	"sync"
+
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/hive.go/runtime/options"
+	"github.com/iotaledger/hive.go/serializer/v2/marshalutil"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+type EpochBasedProvider struct {
+	mutex                             sync.RWMutex
+	protocolParametersByVersion       map[iotago.Version]iotago.ProtocolParameters
+	futureProtocolParametersByVersion map[iotago.Version]iotago.Identifier
+	protocolVersions                  *ProtocolEpochVersions
+
+	latestVersionMutex sync.RWMutex
+	latestVersion      iotago.Version
+
+	currentSlotMutex sync.RWMutex
+	currentSlot      iotago.SlotIndex
+
+	optsAPIForMissingVersionCallback func(version iotago.Version) (iotago.API, error)
+}
+
+func NewEpochBasedProvider(opts ...options.Option[EpochBasedProvider]) *EpochBasedProvider {
+	return options.Apply(&EpochBasedProvider{
+		protocolParametersByVersion:       make(map[iotago.Version]iotago.ProtocolParameters),
+		futureProtocolParametersByVersion: make(map[iotago.Version]iotago.Identifier),
+		protocolVersions:                  NewProtocolEpochVersions(),
+	}, opts)
+}
+
+func (e *EpochBasedProvider) SetCurrentSlot(slot iotago.SlotIndex) {
+	e.currentSlotMutex.Lock()
+	defer e.currentSlotMutex.Unlock()
+
+	e.currentSlot = slot
+}
+
+func (e *EpochBasedProvider) AddProtocolParametersAtEpoch(protocolParameters iotago.ProtocolParameters, epoch iotago.EpochIndex) {
+	e.AddProtocolParameters(protocolParameters)
+	e.AddVersion(protocolParameters.Version(), epoch)
+}
+
+func (e *EpochBasedProvider) AddProtocolParameters(protocolParameters iotago.ProtocolParameters) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
```

Comment:
I think that can actually deadlock, because if someone tries to read the Version in parallel, the latestVersionMutex might be readlocked, but can't be unlocked because it waits for the mutex readlock to be aquired, which is locked here.
So better don't use defer, but unlock after the delete.